TANF Section 3 Parsing/Validation

elipe17 opened the pull request:
## Summary of Changes
- Updated to support parsing and validation of T6 records
Pull request closes #1110 

## How to Test
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```
- Log in as a data analyst.
- Submit `ADS.E2J.FTP3.TS06` as aggregate data.
- Verify that three T6 records have been generated with zero errors.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] section 3 parsed
+ [x] section 3 validated
+ [x] section 3 restructured from quarter level to month level based on cal quarter
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- Updated to support parsing section 3 data

elipe17 committed:
- Added validators and tests for T6 record

elipe17 committed:
- resolve lint errors

elipe17 committed:
- Quick rename

elipe17 committed:
- Update fields in test

jtimpe reviewed (approved):
Looks good!

jtimpe commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py:
```diff
@@ -0,0 +1,168 @@
+              required=True, validators=[validators.month_year_yearIsLargerThan(1998)]),
+        Field(item="3B", name='CALENDAR_QUARTER', type='number', startIndex=6, endIndex=7,
+              required=True, validators=[validators.isInLimits(1, 4)]),
+        Field(item="4C", name='NUM_APPLICATIONS', type='number', startIndex=23, endIndex=31,
```
just a question about the item numbers - we use the `A`, `B`, etc. to denote sub-fields in other schemas, here it's used to denote which record of the row is being processed (corresponding to the "first month", "second month", etc. in the instructions). i thought this was fine and made sense until i noticed `CALENDAR_YEAR` and `CALENDAR_QUARTER` above use it as a sub-field indicator. is it confusing to mix them up in the same schema? could we use a different month indicator?

elipe17 replied:
I think this still makes sense. Even though we internally separate out the record by month, the PDF schema still lists those fields as sub-fields. Also, it isn't immediately clear to me how we would re-assign a value to 3A/3B. I guess we could assign them both as 3 but I don't feel real good about duplicate values in the schema. What are your thoughts?

jtimpe replied:
i was thinking more along the lines of `4-1` in the first schema, `4-2` in the second, and so on. 3A/3B makes sense to me still and that's the way we use it in other schemas

raftmsohani reviewed (approved):
Tests pass locally, all good

andrew-jameson commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py:
```diff
@@ -0,0 +1,168 @@
+        Field(item="3B", name='CALENDAR_QUARTER', type='number', startIndex=6, endIndex=7,
+              required=True, validators=[validators.isInLimits(1, 4)]),
+        Field(item="4A", name='NUM_APPLICATIONS', type='number', startIndex=7, endIndex=15,
+              required=True, validators=[validators.isInLimits(0, 99999999)]),
```
While this might be technically permissible to have all these 9**** field validations, is there a cleaner solution? I just don't love seeing all these hard-coded and sort of absurd "magic number" values. Perhaps we just derive length and push that down to `isInLimits()`? Like, common sense says to me there won't be 99 million applications per quarter or esp. the assistance value being so high...

ADPennington commented:
@elipe17 can you update branch and add qasp label ? 😄

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1110-tanf-section3

elipe17 committed:
- Fixed conflict

elipe17 committed:
Merge branch '1109-tanf-section2-validation' into 1110-tanf-section3

elipe17 committed:
- Fix error from merge conflict

ADPennington commented:
@elipe17 can you look into this? can't see the T6 parsing in DAC. deployed to qasp env. 
![2649p2](https://github.com/raft-tech/TANF-app/assets/63075587/d504ef51-9d43-458f-aba3-5c280796c627)
![2649p1](https://github.com/raft-tech/TANF-app/assets/63075587/4b69f57d-18fc-4d2b-b8a6-f0dc3f8ec835)

elipe17 commented:
> @elipe17 can you look into this? can't see the T6 parsing in DAC. deployed to qasp env. ![2649p2](https://user-images.githubusercontent.com/63075587/261074519-d504ef51-9d43-458f-aba3-5c280796c627.PNG) ![2649p1](https://user-images.githubusercontent.com/63075587/261074521-4b69f57d-18fc-4d2b-b8a6-f0dc3f8ec835.PNG)

@ADPennington  Issue should be resolved. Please give it a shot at your convenience!

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py:
```diff
@@ -0,0 +1,168 @@
+    ],
+)
+
+s3 = RowSchema(
```
@elipe17 one immediate issue i see here (which i think is due at least in part to [this incomplete example](https://user-images.githubusercontent.com/63075587/257940006-b08aeda4-3b62-4580-86cf-fe0f1dd60886.PNG) i added in the ticket), is that its not possible to tell which month is associated with each record. 

_DAC_
![2649p3](https://github.com/raft-tech/TANF-app/assets/63075587/ca4ab10f-3d2e-4bdb-b986-5675a53dca27)

_notebook_
![2649p4](https://github.com/raft-tech/TANF-app/assets/63075587/2335adf9-5507-4bde-8db4-6d2c74a5d130)

ADPennington commented on tdrs-backend/tdpservice/parsers/schema_defs/tanf/t6.py:
```diff
@@ -0,0 +1,168 @@
+    fields=[
+        Field(item="0", name='RecordType', type='string', startIndex=0, endIndex=2,
+              required=True, validators=[]),
+        Field(item="3A", name='CALENDAR_YEAR', type='number', startIndex=2, endIndex=6,
```
@elipe17 do we need this? this is not a data element. `CALENDAR_QUARTER` has a poor label by default unfortunately, but its technically exactly the same as the label in the header and takes on value: YYYYQ. I dont feel too strongly about this;  most important is to add `RPT_MONTH_YEAR` based on YYYYQ when restructuring to month-level records.

elipe17 committed:
- Updated to create RPT_MONTH_YEAR from CALENDAR_QUARTER
- Updated tests

elipe17 committed:
- Added new validators for field change

elipe17 committed:
- Remove debug code

elipe17 committed:
- Genericize TransformField
- Update all schemas to use TransformField
- Move transforms to seperate file
- Fix lint errors

elipe17 committed:
- put kwargs in correct spot

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 1110-tanf-section3

elipe17 committed:
- Very minor change to avoid setting field unnecessarily on all TransformFields

ADPennington commented:
> > @elipe17 can you look into this? can't see the T6 parsing in DAC. deployed to qasp env. ![2649p2](https://user-images.githubusercontent.com/63075587/261074519-d504ef51-9d43-458f-aba3-5c280796c627.PNG) ![2649p1](https://user-images.githubusercontent.com/63075587/261074521-4b69f57d-18fc-4d2b-b8a6-f0dc3f8ec835.PNG)
> 
> @ADPennington Issue should be resolved. Please give it a shot at your convenience!

@elipe17 this issue has returned in qasp env. can you check into db?

elipe17 commented:
@ADPennington I resolved the issue. You should be good to test at your convenience.

ADPennington commented on tdrs-backend/tdpservice/parsers/validators.py:
```diff
@@ -51,6 +51,19 @@ def if_then_validator_func(value):
+
+        condition_val = value[condition_field] if type(value) is dict else getattr(value, condition_field)
+        return (True, None) if sum == condition_val else (False,
+                                                          f"The sum of {sum_fields} does not equal {condition_val}.")
```
@elipe17 is it possible to change `condition_val` to the data element label? the value is not as meaningful. 

![s3_val](https://github.com/raft-tech/TANF-app/assets/63075587/1b0c2c92-513d-432f-a212-052e1986cad9)

elipe17 committed:
- Updated validator to call out name of field versus the value
- Updated tests

elipe17 committed:
- Fix lint

elipe17 committed:
Merge branch '1109-tanf-section2-validation' of https://github.com/raft-tech/TANF-app into 1110-tanf-section3

ADPennington reviewed (approved):
looks great! thanks @elipe17 🚀
